klippa/src/serialize.rs

```diff
@@ -283,6 +314,125 @@ impl Serializer {
         Some(obj_idx)
     }
 
+    pub(crate) fn pop_discard(&mut self) {
+        if self.current.is_none() {
+            return;
+        }
+        // Allow cleanup when we've error'd out on int overflows which don't compromise the serializer state.
+        if self.in_error() && !self.only_overflow() {
+            return;
+        }
+
+        let pool_idx = self.current.unwrap();
+        let Some(obj) = self.object_pool.get_obj(pool_idx) else {
+            return;
+        };
+        self.current = obj.next_obj;
+        //TODO: consider zerocopy
+        self.revert(obj.head, obj.tail);
+        self.object_pool.release(pool_idx);
+    }
+
+    fn revert(&mut self, snap_head: usize, snap_tail: usize) {
+        if self.in_error() || self.head < snap_head || self.tail > snap_tail {
+            return;
+        }
+        self.head = snap_head;
+        self.tail = snap_tail;
+        self.discard_stale_objects();
+    }
+
+    fn discard_stale_objects(&mut self) {
+        if self.in_error() {
+            return;
+        }
+
+        let len = self.packed.len();
+        for i in (0..len).rev() {
+            let pool_idx = self.packed[i];
+            let Some(obj) = self.object_pool.get_obj(pool_idx) else {
+                self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                return;
+            };
+
+            if obj.next_obj.is_some() {
+                self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                return;
+            }
+
+            if obj.head >= self.tail {
+                break;
+            }
+
+            let hash = hash_one_pool_idx(pool_idx, &self.data, &self.object_pool);
+            self.packed_map
+                .del(pool_idx, hash, &self.data, &self.object_pool);
+            self.object_pool.release(pool_idx);
+            self.packed.pop();
+        }
+
+        if let Some(pool_idx) = self.packed.last() {
+            let Some(obj) = self.object_pool.get_obj(*pool_idx) else {
+                self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+                return;
+            };
+
+            if obj.head != self.tail {
+                self.set_err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+            }
+        }
+    }
+
+    pub(crate) fn add_link(
+        &mut self,
+        offset_byte_range: Range<usize>,
+        obj_idx: ObjIdx,
+        whence: OffsetWhence,
+        bias: u32,
+        is_signed: bool,
+    ) {
+        if self.in_error() || self.current.is_none() {
+            return;
+        }
+
+        let pool_idx = self.current.unwrap();
+        let Some(current) = self.object_pool.get_obj_mut(pool_idx) else {
+            return;
+        };
+
+        let Some(link_width) = LinkWidth::new_checked(offset_byte_range.len()) else {
```

Comment:
In the error case here we shouldn't silently ignore the invalid width.  If we have a caller accidentally pass in a range of an unsupported width it will seem like everything worked which will possibly lead to an unexpected bug in the produced serialization. Recommend either:
1. set the serializer into an error state.
2. add Result<(), ...> as the return type to signal the failure back to the caller.
3. or add an assert to cause a panic.

I prefer option 1 or 2.